src/04-spin-boson-model.md

```diff
@@ -0,0 +1,776 @@
+
+```python
+def make_operators(max_bosons, parity):
+    
+    a  = tensor(destroy(max_bosons+1), qeye(2))     # tensorised boson destruction operator
+    sx = tensor(qeye(max_bosons+1), sigmax())       # tensorised sigma_x operator
+    sz = tensor(qeye(max_bosons+1),sigmaz())        # tensorised sigma_z operator
+    
+    two_state     =  1/2*sz                         # two state system energy operator
+    bosons       =  (a.dag()*a+0.5)                 # boson energy operator
+    number        = a.dag()*a                       # boson number operator
+    interaction  = (a.dag() + a) * sx               # interaction energy operator      
+    
+    P = sz*(1j*np.pi*a.dag()*a).expm()              # parity operator 
+    
+    # map from QuTiP number states to |n,±> states
+    possible_ns = range(0, max_bosons+1)
+    possible_ms = ["+","-"]
+    nm_list = [(n,m) for (n,m) in product(possible_ns, possible_ms)]
+    
+    # only do parity extraction if a valid parity is being used
+    if (parity==1) | (parity==-1):
+        p           = np.where(P.diag()==parity)[0]
+        
+        two_state   = two_state.extract_states(p)
+        bosons      = bosons.extract_states(p)
+        number      = number.extract_states(p)
+        interaction = interaction.extract_states(p)
+        nm_list     = [nm_list[i] for i in p]
+    
+    
+    return two_state, bosons, interaction, number, nm_list
+```
+
+We are now in a position to compare the energy level plots for the different parity universes side by side.
+
+```python
+# ODD PARITY
+
+two_state, bosons, interaction, number, nm_list = make_operators(max_bosons=4, parity=-1)
+
+df_odd = make_df_for_energy_scan("$\Delta E$", -4, 4, 201, two_state.shape[0])
+
+for i, row in df_odd.iterrows():
+    H =  row["$\Delta E$"]*two_state + 1*bosons + 0.2*interaction
+    evals, ekets = H.eigenstates()
+    df_odd.iloc[i,1:] = evals 
+```
+
+```python
+# EVEN PARITY
+
+two_state, bosons, interaction, number, nm_list = make_operators(max_bosons=4, parity=1)
+
+df_even = make_df_for_energy_scan("$\Delta E$", -4, 4, 201, two_state.shape[0])
+
+for i, row in df_even.iterrows():
+    H =  row["$\Delta E$"]*two_state + 1*bosons + 0.2*interaction
+    evals, ekets = H.eigenstates()
+    df_even.iloc[i,1:] = evals 
+```
+
+```python
+fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(15,6), sharey=True)
+
+
+df_odd.plot(x="$\Delta E$",ylim=[-0.5,5.5],legend=False, 
+        title="Odd stationary states ($\omega=1$, $U=0.2$)     (Fig 9)",  ax=axes[0]);
+
+df_even.plot(x="$\Delta E$",ylim=[-0.5,5.5],legend=False, 
+        title="Even stationary states ($\omega=1$, $U=0.2$)     (Fig 10)",  ax=axes[1]);
+
+axes[0].set_ylabel("Energy");
+```
+
+Figs 9 and 10 show us the energy of the stationary states for the odd and even parity states respectively. As we expected, we now only see anti-crossings because in each universe there are no forbidden interactions.
```

Comment:
I would add:" (please now notice that all lines anticross, also at DeltaE=3Omega, like Fig 3)"

Or repeat a zoom, but that maybe is redundant



It might also be very GOOD to plot here the labels for the lines which, as opposed to Fig 1, now don't have to say level_, but they can say |0,+> etc!

Maybe a note on this here reminding why we changed it compare to Fig 1-3 may help

What do you think?

Reply:
I'm not sure we need to reference the zoom of figure 3 again. I will, however, reword the sentence to try and make things clearer.

Regarding line labels. I agree that we don't need the level number labels here. I'm also not sure I agree with putting state labels like |0,+>. I don't think that |n,±> state is preserved along a line. e.g. in fig 10, after the orange line anti-crosses with the green, I don't think it is in the same |n,±> state - the orange and green kind of swap identities. I don't explicitly show it in the notebook, but it might be worth exploring in another notebook, e.g. we could actually demonstrate the [Landau–Zener effect](https://en.wikipedia.org/wiki/Landau%E2%80%93Zener_formula)

Re a note as to why we changed from having all parities to only a single one - it is worth reiterating that. Tbh, we don't technically need to do it in order to do the down conversation simulation but it gets us unto a good practice or first selecting the parity universe and them simulating